utils/dir/dir.go

```diff
@@ -0,0 +1,84 @@
+// pmm-managed
+// Copyright (C) 2017 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+package dir
+
+import (
+	"fmt"
+	"os"
+	"os/user"
+	"strconv"
+	"syscall"
+)
+
+// Params represent the input for CreateDataDir
+type Params struct {
+	Path  string
+	Perm  os.FileMode
+	Chown bool
+	User  string
+	Group string
+}
+
+// CreateDataDir creates/updates directories with the given permissions in the persistent volume.
+func CreateDataDir(params Params) error {
```

Comment:
Let's change that function to return the first encountered error (mostly for tests), but continue as far as possible with its flow. For example, if chmod can't be done, store that error (if it is the first one), but still continue with chown.

And where are the unit tests?

Reply:
changed it to store the error and return the first encountered one everywhere except for three places where I am still returning the error right away
- when we create dir as there is no point in continuing to chmod and chown if the dir wasn't created
- when we do user lookup because if we continue despite error it can panic
- group lookup same reason as above

Reply:
i tried to make chmod fail with invalid permissions like `77777` to check for the first encountered error but it didn't fail somehow the std lib is converting invalid permissions to valid ones. not sure how else to test for it

Reply:
>  For example, if chmod can't be done, store that error (if it is the first one), but still continue with chown.

But we don't continue in the current code – we return stored error

Reply:
if the chmod error is the first encountered error and there are no errors between the chmod and chown call then we can chown even if chmod wasn't successful, I thought thats what you meant by returning the first encountered error. so I added a check for it everywhere.